### PR TITLE
add test_replica_zone_anti_affinity

### PR DESCRIPTION
**Pull request by meldafrawi**
longhorn/longhorn#1052

Note:
 - I've moved `test_replica_zone_anti_affinity` to `test_zone.py`
Signed-off-by: Mohamed Eldafrawi <mohamed.eldafrawi@rancher.com>

### CODE REVIEW

File: manager/integration/tests/test_zone.py [hunk 1]
```diff
@@ -14,8 +14,14 @@
 from random import randrange
 from test_scheduling import wait_new_replica_ready
 from common import get_version_api_client
+from common import CONDITION_STATUS_TRUE
+from common import wait_for_volume_condition_scheduled
+from common import wait_for_volume_delete
+from test_scheduling import SETTING_REPLICA_SOFT_ANTI_AFFINITY
 
 
+SETTING_ZONE_ANTI_AFFINITY = "replica-zone-soft-anti-affinity"
```
Comment (boknowswiki) [inline manager/integration/tests/test_zone.py:23]: Better to name the variable `SETTING_ZONE_SOFT_ANTI_AFFINITY`, it's more clear that it's hard or soft.

File: manager/integration/tests/test_zone.py [hunk 2]
```diff
@@ -220,3 +239,78 @@ def test_zone_tags(client, core_api, volume_name, k8s_node_zone_tags):  # NOQA
...
+    12. Clean up the replica count, the zone labels and the volume.
+    """
+
+    wait_longhorn_node_zone_updated(client)
+
+    node_anti_affinity_setting = \
+        client.by_id_setting(SETTING_REPLICA_SOFT_ANTI_AFFINITY)
+    client.update(node_anti_affinity_setting, value="false")
+
+    zone_anti_affinity_setting = \
```
Comment (boknowswiki) [inline manager/integration/tests/test_zone.py:270]: Same here, better with `zone_soft_anti_affinity_setting`

File: manager/integration/tests/test_zone.py [hunk 3]
```diff
@@ -220,3 +239,78 @@ def test_zone_tags(client, core_api, volume_name, k8s_node_zone_tags):  # NOQA
...
+    volume = create_and_check_volume(client, volume_name)
+
+    assert volume.conditions.scheduled.status == "False"
+    assert volume.conditions.scheduled.reason == "ReplicaSchedulingFailure"
+
+    lh_nodes = client.list_node()
+
+    node_name = lh_nodes[-1].name
+
+    set_k8s_node_zone_label(core_api, node_name, "lh-zone3")
```
Comment (boknowswiki) [inline manager/integration/tests/test_zone.py:283]: My understanding here is `k8s_node_zone_tags ` will set 3 nodes with 2 zones. How can we know the order? And can we guarantee we will get 3 nodes in three zones here? Is that possible like this node1 & node 3 in zone 1 and we just set node 2 from zone 2 to zone 3?

File: manager/integration/tests/test_zone.py [hunk 4]
```diff
@@ -220,3 +239,78 @@ def test_zone_tags(client, core_api, volume_name, k8s_node_zone_tags):  # NOQA
...
+
+    wait_longhorn_node_zone_updated(client)
+
+    wait_for_volume_condition_scheduled(client, volume_name,
+                                        "status",
+                                        CONDITION_STATUS_TRUE)
+
+    zone_anti_affinity_setting = \
```
Comment (boknowswiki) [inline manager/integration/tests/test_zone.py:291]: Same here, better with `zone_soft_anti_affinity_setting`

File: manager/integration/tests/test_zone.py (hunk 1 above)
Comment (meldafrawi) [reply manager/integration/tests/test_zone.py]: Fixed, thanks

File: manager/integration/tests/test_zone.py (hunk 3 above)
Comment (meldafrawi) [reply manager/integration/tests/test_zone.py]: Fixed

File: manager/integration/tests/test_zone.py (hunk 4 above)
Comment (meldafrawi) [reply manager/integration/tests/test_zone.py]: fixed

File: manager/integration/tests/test_zone.py (hunk 2 above)
Comment (meldafrawi) [reply manager/integration/tests/test_zone.py]: fixed

File: manager/integration/tests/test_zone.py [hunk 5]
```diff
@@ -14,8 +14,14 @@
...
+from common import CONDITION_STATUS_TRUE
+from common import wait_for_volume_condition_scheduled
+from common import wait_for_volume_delete
+from test_scheduling import SETTING_REPLICA_SOFT_ANTI_AFFINITY
 
 
+SETTING_ZONE_SOFT_ANTI_AFFINITY = "replica-zone-soft-anti-affinity"
+
 # label deprecated for k8s >= v1.17
 DEPRECATED_K8S_ZONE_LABEL = "failure-domain.beta.kubernetes.io/zone"
```
Comment (boknowswiki) [inline manager/integration/tests/test_zone.py:26]: I think for k8s >= v1.17, the label is `topology.kubernetes.io/zone`, and for k8s <v1.17, it's `failure-domain.beta.kubernetes.io/zone`.

File: manager/integration/tests/test_zone.py [hunk 6]
```diff
@@ -220,3 +239,79 @@ def test_zone_tags(client, core_api, volume_name, k8s_node_zone_tags):  # NOQA
...
+    8. Set zone anti-affinity to soft.
+    9. Change the zone labels on node 1 & 2 & 3 to "zone1".
+    10. Create a volume.
+    11. Wait for volume condition `scheduled` to be success.
+    12. Clean up the replica count, the zone labels and the volume.
+    """
+
+    wait_longhorn_node_zone_updated(client)
+
+    node_anti_affinity_setting = \
```
Comment (boknowswiki) [inline manager/integration/tests/test_zone.py:266]: maybe `node_soft_anti_affinity_setting ` is better?

File: manager/integration/tests/test_zone.py (hunk 5 above)
Comment (meldafrawi) [reply manager/integration/tests/test_zone.py:26]: Yes, you're totally right, we've this function to return the correct label based on kubernetes version

https://github.com/longhorn/longhorn-tests/blob/f70218fad05fafc1f3fc59f144f631b44b18d6f3/manager/integration/tests/test_zone.py#L34

File: manager/integration/tests/test_zone.py [hunk 7]
```diff
@@ -220,3 +239,79 @@ def test_zone_tags(client, core_api, volume_name, k8s_node_zone_tags):  # NOQA
...
+
+    volume = create_and_check_volume(client, volume_name)
+    wait_for_volume_condition_scheduled(client, volume_name,
+                                        "status",
+                                        CONDITION_STATUS_TRUE)
+
+    # reset default setting
+    node_anti_affinity_setting = \
+        client.by_id_setting(SETTING_REPLICA_SOFT_ANTI_AFFINITY)
+    client.update(node_anti_affinity_setting, value="true")
```
Comment (boknowswiki) [inline manager/integration/tests/test_zone.py:313]: With this change [1282](https://github.com/longhorn/longhorn/issues/1282), the node soft anti-affinity will be fasle soon.

Comment (meldafrawi) [reply manager/integration/tests/test_zone.py]: updated

**Comment by yasker** [comment]
LGTM.

**Pull request merged**